host/sgx/create.c

```diff
@@ -588,58 +621,63 @@ oe_result_t oe_sgx_validate_enclave_properties(
-static oe_result_t _add_eeid_marker_page(
+static oe_result_t _add_eeid_page(
     oe_sgx_load_context_t* context,
     oe_enclave_t* enclave,
-    size_t image_size,
     uint64_t entry_point,
     oe_sgx_enclave_properties_t* props,
     uint64_t* vaddr)
 {
     oe_result_t result = OE_UNEXPECTED;
     oe_eeid_t* eeid = context->eeid;
 
-    if (eeid && is_eeid_base_image(props) &&
-        context->type == OE_SGX_LOAD_TYPE_CREATE)
+    oe_page_t* page = oe_memalign(OE_PAGE_SIZE, sizeof(oe_page_t));
+    memset(page, 0, sizeof(oe_page_t));
+
+    if (eeid)
     {
-        // Finalize the memory settings
-        props->header.size_settings = eeid->size_settings;
+        /* Finalize the memory settings */
+        if (is_eeid_base_image(&props->header.size_settings))
+            props->header.size_settings = eeid->size_settings;
+        else
+            eeid->size_settings = props->header.size_settings;
+
+        /* EEID must fit on a single page. */
+        size_t eeid_page_size = sizeof(oe_eeid_page_t) + eeid->signature_size;
+        if (eeid_page_size > OE_PAGE_SIZE)
+            return OE_INVALID_PARAMETER;
 
-        // Record EEID information
+        /* Record EEID information */
         eeid->version = OE_EEID_VERSION;
         oe_sha256_context_t* hctx = &context->hash_context;
         oe_sha256_save(hctx, eeid->hash_state.H, eeid->hash_state.N);
         eeid->entry_point = entry_point;
         eeid->vaddr = *vaddr;
         eeid->signature_size = sizeof(sgx_sigstruct_t);
-        memcpy(
-            eeid->data + eeid->data_size,
-            (uint8_t*)&props->sigstruct,
-            sizeof(sgx_sigstruct_t));
-
-        oe_page_t* page = oe_memalign(OE_PAGE_SIZE, sizeof(oe_page_t));
-        memset(page, 0, sizeof(oe_page_t));
-        oe_eeid_marker_t* marker = (oe_eeid_marker_t*)page;
+        memcpy(eeid->signature, props->sigstruct, eeid->signature_size);
 
-        /* The offset to the EEID in marker->offset is also the extended
-         * commit size of the base image and dynamically configured data
-         * pages (stacks + heap) excluding the EEID data size.
-         */
-        _calculate_enclave_size(image_size, props, &marker->offset, NULL);
+        /* Compose EEID memory page */
+        oe_eeid_page_t* eeid_page = (oe_eeid_page_t*)page;
+        memcpy(
+            &eeid_page->eeid, eeid, sizeof(oe_eeid_t) + eeid->signature_size);
 
-        uint64_t addr = enclave->addr + *vaddr;
-        uint64_t src = (uint64_t)page;
-        uint64_t flags = SGX_SECINFO_REG | SGX_SECINFO_R | SGX_SECINFO_W;
+        /* Compute config hash */
+        oe_sha256(
+            context->config.data, context->config.size, &eeid_page->config_id);
+    }
+    else
+    {
+        /* Add an empty/guard page */
+    }
 
-        OE_CHECK(oe_sgx_load_enclave_data(
-            context, enclave->addr, addr, src, flags, false));
-        (*vaddr) += OE_PAGE_SIZE;
-        free(page);
+    uint64_t addr = enclave->addr + *vaddr;
```

Comment:
What do you mean by "committed"? It currently adds a measured page of all zeroes to non-EEID enclaves and the code relies on that to distinguish between non-EEID (0) and EEID (non-0).

Reply:
"not committed" means the linear address does not map to any EPC page, and access to the linear address will cause page fault. See https://github.com/bodzhang/openenclave/blob/EEIDDesignChange/docs/DesignDocs/ExtendedEnclaveInitializationData_V2.md for detail. 
For a regualr enclave, the implementaiton shoudl not add an all-0 page, instead, the _add_eeid_page(..) operation shoudl be skipped.

Reply:
What would be the benefit of that? I currently use the fact that it's zero to recognize the fact that there is no EEID, which is very convenient. If we can't rely on this page being initialized, then I will need another new page that's initialized (to house that bit of information) or a custom flag in base images that's set if an image is intended only for EEID use.

Reply:
Adding a gaurd page instead of all-0 page between the Code and Initialized Data pages and the heap pages in the regualr enclave, the design causes minimum or argurably no impact to the regular enclave. Guard pages can't be accessed by enclave code, even if some code bug allows the attacker to trigger the enclave code to access memeory it's not suppoed to access, and are already used to seperate different sections of enclave pages in the existing regualr enclave design. With regard to the mechanism to distinguish EEID enclave for regualr enclave, I think an explict flag, for example, a EEID_SGX_Enclaev type in oe_enclave_type_t, is a cleaner design, with better readability and lesser security analysis need.

Reply:
There is now a clean setting (i.e. `oe_enclave_setting_eeid_t`) that steers the logic and non-EEID images neither have a gap in the address space nor a measured page, so no impact on existing non-EEID images.